odoo: fix odoo16; add odoo17

siriobalmelli opened the pull request:
Preserve 'odoo16' on the same pattern as existing 'odoo15'

## Description of changes

- updated Odoo version
- added self as a maintainer

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ N/A ] x86_64-darwin
  - [ N/A ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

- tested updated python textfixtures package https://github.com/NixOS/nixpkgs/pull/309637

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pluiedev commented on pkgs/applications/finance/odoo/odoo16.nix:
```diff
@@ -0,0 +1,115 @@
+  doCheck = false;
+
+  makeWrapperArgs = [
+    "--prefix" "PATH" ":" "${lib.makeBinPath [ wkhtmltopdf rtlcss ]}"
```
[`wkhtmltopdf` depends on `qtwebkit`, which is long unmaintained and marked insecure.](https://github.com/NixOS/nixpkgs/issues/297646) Is it absolutely certain oboo still needs this? Because at the moment, Nixpkgs refuses to evaluate it without enabling insecure packages

siriobalmelli replied:
Good question! I'll look into this.

siriobalmelli replied:
I've gone through the source and the news is not good here: `wkhtmltopdf` is used by builtin `i18n` and `ir` modules:

- it's still in use
- it can't be made an optional feature

siriobalmelli commented:
Respun to:

- fix existing broken odoo16 build before migration
- move to pkgs/by-name structure so as to pass pre-merge cheks

siriobalmelli commented:
Rebased on top of latest `master`.

@mkg20001 I've polished this as much as I know how, please review.

siriobalmelli committed:
odoo: fix build

Pinned werkzeug override fails a test; disable it.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

siriobalmelli committed:
odoo: 16.0.20231024 -> 17.0.20240507

Preserve 'odoo16' on the same pattern as existing 'odoo15'

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

siriobalmelli committed:
odoo: migrate to pkgs/by-name

Remove original 'pkgs/applications/finance/odoo' directory.

Update 'pkgs/applications/finance/odoo/update.sh'
to point to the nightly builds archive for odoo, ignoring Docker.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

siriobalmelli committed:
odoo: 17.0.20240507 -> 17.0.20240616

Make it possible to override 'odoo_version' and 'odoo_release'
by moving them to the input attribute set with default values.

Modify 'update.sh' so it correctly updates these defaults.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

siriobalmelli committed:
tests/odoo: fix broken test

Odoo does not initialize the database by default,
it must be explicitly initialized for the test as written to succeed.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

siriobalmelli commented:
> Odoo (nixosTests.odoo) test fails with
> 
> ```
> server # [   30.340202] odoo-start[973]: 2024-06-17 12:24:56,053 973 ERROR odoo odoo.http: Exception during request handling.
> server # [   30.341666] odoo-start[973]: Traceback (most recent call last):
> server # [   30.342585] odoo-start[973]:   File "/nix/store/lp8xszwz00h3mfi7nnfx11as3nmsj0xd-odoo-17.0.20240610/lib/python3.10/site-packages/odoo/http.py", line 1748, in _serve_db
> server # [   30.344536] odoo-start[973]:     return service_model.retrying(self._serve_ir_http, self.env)
> server # [   30.345641] odoo-start[973]:   File "/nix/store/lp8xszwz00h3mfi7nnfx11as3nmsj0xd-odoo-17.0.20240610/lib/python3.10/site-packages/odoo/service/model.py", line 133, in retrying
> server # [   30.347604] odoo-start[973]:     result = func()
> server # [   30.349071] odoo-start[973]:   File "/nix/store/lp8xszwz00h3mfi7nnfx11as3nmsj0xd-odoo-17.0.20240610/lib/python3.10/site-packages/odoo/http.py", line 1760, in _serve_ir_http
> server # [   30.351416] odoo-start[973]:     ir_http = self.registry['ir.http']
> server # [   30.352399] odoo-start[973]:   File "/nix/store/lp8xszwz00h3mfi7nnfx11as3nmsj0xd-odoo-17.0.20240610/lib/python3.10/site-packages/odoo/modules/registry.py", line 213, in __getitem__
> server # [   30.355061] odoo-start[973]:     return self.models[model_name]
> server # [   30.355989] odoo-start[973]: KeyError: 'ir.http'
> ```

Thanks @mkg20001, pushed a fix for the test.

I notice 2 issues with the odoo service module as-is:

1. Database initialization needed manually on a new db, as shown in the fix for the test.
1. Assumes local postgres: no option to connect to an external db.
1. No option to pass a custom configuration path: any secrets (eg: db secrets) need to be in the clear; passing a configuration path would allow solutions such as [sops-nix](https://github.com/Mic92/sops-nix).

I'm curious to hear your thoughts on the service module and where you see it evolving? If it matches what I'm working on I'd be happy to contribute something as a follow-up PR 😄

mkg20001 commented:
3. we can add a secretsFile that is merged at service start with the other
options, like a few modules do
2. we can have it optionally, but keep local db as default
1. not sure as we'd be automatically running migrations?

I'll run the tests then merge this pr, you can make a follow-up pr with
those changes.

Sirio Balmelli ***@***.***> schrieb am Mi., 19. Juni 2024,
17:43:

> Odoo (nixosTests.odoo) test fails with
>
> server # [   30.340202] odoo-start[973]: 2024-06-17 12:24:56,053 973 ERROR odoo odoo.http: Exception during request handling.
> server # [   30.341666] odoo-start[973]: Traceback (most recent call last):
> server # [   30.342585] odoo-start[973]:   File "/nix/store/lp8xszwz00h3mfi7nnfx11as3nmsj0xd-odoo-17.0.20240610/lib/python3.10/site-packages/odoo/http.py", line 1748, in _serve_db
> server # [   30.344536] odoo-start[973]:     return service_model.retrying(self._serve_ir_http, self.env)
> server # [   30.345641] odoo-start[973]:   File "/nix/store/lp8xszwz00h3mfi7nnfx11as3nmsj0xd-odoo-17.0.20240610/lib/python3.10/site-packages/odoo/service/model.py", line 133, in retrying
> server # [   30.347604] odoo-start[973]:     result = func()
> server # [   30.349071] odoo-start[973]:   File "/nix/store/lp8xszwz00h3mfi7nnfx11as3nmsj0xd-odoo-17.0.20240610/lib/python3.10/site-packages/odoo/http.py", line 1760, in _serve_ir_http
> server # [   30.351416] odoo-start[973]:     ir_http = self.registry['ir.http']
> server # [   30.352399] odoo-start[973]:   File "/nix/store/lp8xszwz00h3mfi7nnfx11as3nmsj0xd-odoo-17.0.20240610/lib/python3.10/site-packages/odoo/modules/registry.py", line 213, in __getitem__
> server # [   30.355061] odoo-start[973]:     return self.models[model_name]
> server # [   30.355989] odoo-start[973]: KeyError: 'ir.http'
>
> Thanks @mkg20001 <https://github.com/mkg20001>, pushed a fix for the test.
>
> I notice 2 issues with the odoo service module as-is:
>
>    1. Database initialization needed manually on a new db, as shown in
>    the fix for the test.
>    2. Assumes local postgres: no option to connect to an external db.
>    3. No option to pass a custom configuration path: any secrets (eg: db
>    secrets) need to be in the clear; passing a configuration path would allow
>    solutions such as sops-nix <https://github.com/Mic92/sops-nix>.
>
> I'm curious to hear your thoughts on the service module and where you see
> it evolving? If it matches what I'm working on I'd be happy to contribute
> something as a follow-up PR 😄
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/312060#issuecomment-2179009168>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3AO2NXOQT54CCUTFHUZ3LZIGRJZAVCNFSM6AAAAABHZBBBB2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNZZGAYDSMJWHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
